rpc: allow writing UTXO set to a named pipe

theStack opened the pull request:
This PR slightly modifies the `dumptxoutset` RPC to allow writing the UTXO set dump into a [named pipe](https://askubuntu.com/a/449192), so that the output data can be consumed by another process, see #31373. Taking use of this with the utxo-to-sqlite.py tool (introduced in #27432), creating an UTXO set in SQLite3 format is possible on the fly. E.g. for signet:
```
$ mkfifo /tmp/utxo_fifo && ./build/bin/bitcoin-cli -signet dumptxoutset /tmp/utxo_fifo latest &
$ ./contrib/utxo-tools/utxo_to_sqlite.py /tmp/utxo_fifo ./utxo.sqlite
UTXO Snapshot for Signet at block hash 000000012711f0a4e741be4a22792982..., contains 61848352 coins
1048576 coins converted [1.70%], 2.800s passed since start
....
....
60817408 coins converted [98.33%], 159.598s passed since start
{
  "coins_written": 61848352,
  "base_hash": "000000012711f0a4e741be4a22792982370f51326db20fca955c7d45da97f768",
  "base_height": 294305,
  "path": "/tmp/utxo_fifo",
  "txoutset_hash": "34ae7fe7af33f58d4b83e00ecfc3b9605d927f154e7a94401226922f8e3f534e",
  "nchaintx": 28760852
}
TOTAL: 61848352 coins written to ./utxo.sqlite, snapshot height is 294305.
```
Note that the `dumptxoutset` RPC calculates an UTXO set hash as a first step before any data is emitted, so especially on mainnet it takes quite a while until the conversion starts and something is happening visibly.

theStack committed:
rpc: support writing UTXO set dump (`dumptxoutset`) to a named pipe

This allows external tooling (e.g. converters) to consume the output
directly, rather than having to write the dump to disk first and then
read it from there again.

Co-authored-by: Luke Dashjr <luke-jr+git@utopios.org>

theStack committed:
test: add test for utxo-to-sqlite conversion using named pipe

tdb3 reviewed (commented):
Approach ACK

Great feature!

Did some manual santiy testing on mainnet:
 - Used `dumptxoutset` to create a dump (with a node synced to block 876,186), `utxo_to_sqlite.py` to covert to a sqlite file, and opened/parsed in python. Conversion seemed successful, the correct number of coins were present in the table
 - Used `dump_to_sqlite.sh` to do the same with fifo (but with a node synced to block 200,000), then open/parsed in python. Conversion seemed successful, the correct number of coins were present in the table

Left a few relatively small comments.
May circle back and review `utxo_to_sqlite.py` in more detail as time allows.

romanz commented on test/functional/tool_utxo_to_sqlite.py:
```diff
@@ -0,0 +1,131 @@
+"""Test utxo-to-sqlite conversion tool"""
+import os
+try:
+    import sqlite3
```
Is the `import` expected to fail on CI?
If so, I am not sure that the test below can run successfully if `sqlite3` is not available...

tdb3 reviewed (approved):
ACK 59df8480be77a8e3618c9422536c4f8aed82467e

Also re-ran tests in https://github.com/bitcoin/bitcoin/pull/31560#pullrequestreview-2522042088

luke-jr commented:
Can we make the exists/is_fifo/open atomic somehow? Seems liable to have a race here someday...

theuni commented:
Not opposed to this, but it seems like a good use-case for a kernel util :)

@thecharlatan @stickies-v @josibake

luke-jr commented:
eg https://github.com/luke-jr/bitcoin/commit/56ee485533820d8c7f0e4e27acd712aa8b411286

virtu commented:
ACK 4c8e9b4f35be4104002ece15b6f72c360756f5d9

Tested on Linux and MacOS.

Nice work. I'm looking to integrating some UTXO sats on my website, so it's nice to see people working on conveniently extracting this data.

yancyribbens commented:
Interesting - I read the link above "name pipe" and it seems the main benefit to a named pipes is just that for performance, one can skip writing to disk and the churn of creating a new file.  I'm not sure I see the perf benefit here in a named pipe though..

yancyribbens commented:
I guess a follow commit is how big is the dump.  If it's very large then that would be a good motivation for a named pipe.

naiyoma reviewed (commented):
Concept ACK

briefly tested on regtest: 
```

./contrib/utxo-tools/dump_to_sqlite.sh "./build/src/bitcoin-cli -regtest" ~/utxos_03_regtest.sqlite3
UTXO Snapshot for Regtest at block hash 1859fa04bf13269212d2cbcc406c91b6..., contains 1358 coins
{
  "coins_written": 1358,
  "base_hash": "1859fa04bf13269212d2cbcc406c91b6d11decbc39b97de79ae75566f5b2e826",
  "base_height": 1338,
  "path": "/tmp/tmp.NbBipQ3mhb/utxos.fifo",
  "txoutset_hash": "3bdfcb5a529096de8e73b516005d4263b7924a6c5f81e0823e9ec904126bbf4d",
  "nchaintx": 1364
}
TOTAL: 1358 coins written to /home/ubuntu/utxos_03_regtest.sqlite3, snapshot height is 1338.
```
 Verified UTXO count matches:
```
sqlite3 ~/utxos_03_regtest.sqlite3 "SELECT COUNT(*) FROM utxos;"
1358
```

theStack commented:
Rebased on master (due to conflict with recently merged #32983, commit b63428ac9ce2c903670409b3e47b9f6730917ae8).

rkrux commented on contrib/utxo-tools/dump_to_sqlite.sh:
In ed02f67c5831d94720e55127216aa0d97f55e72a "contrib: add script dump_to_sqlite.sh for direct SQLite3 UTXO dump"

Nit: Can call the script `dump_utxo_to_sqlite.sh` for clarity even though it resides in `utxo-tools/`.

rkrux reviewed (approved):
lgtm ACK ed02f67c5831d94720e55127216aa0d97f55e72a

sedited reviewed (approved):
ACK 32ecef55a479c98f252b9213732d67d4c5ef8b67

Edit: I'm not fully sold on the shell script. If you interrupt it, you have to manually do some cleanup, otherwise `bitcoind` will not shutdown. Maybe the last commit can just be dropped, seems easy enough to just start bitcoin-cli and the sqlite conversion script in a separate shell.

sedited reviewed (approved):
Re-ACK 509d871fc00e3dd5a9c42e267da13a322b646d76

rkrux reviewed (commented):
Raised one point, rest looks fine.

ajtowns reviewed (commented):
utACK 509d871fc00e3dd5a9c42e267da13a322b646d76

I don't think this PR does "introduce dump_to_sqlite.sh" any more. Is there a way to tell if the streaming data you receive over the pipe is complete or not? Checking `coins_count` from the metadata equals the number of coins you actually received, I guess?

[EDIT: The PR title references the sh script still. I don't think it makes sense to have a shell script to do that; the `utxo_to_sqlite.py` script should just make the fifo and invoke bitcoin-cli itself if we want to make things easier]

theStack commented:
Thanks for the reviews! Removed the shell script mentions from the PR title and the description, and updated the usage example (now being a "two liner", including the pipe creation via `mkfifo`).

> Is there a way to tell if the streaming data you receive over the pipe is complete or not? Checking `coins_count` from the metadata equals the number of coins you actually received, I guess?

Yes, I think that's in fact the only way. Detecting this for a named pipe is not different from an actual file though (which could also be truncated and thus be incomplete): consumers should read for exactly the number of coins specified in the metadata (in the `utxo_to_sqlite.py` script, this is done with a for loop over `num_utxos`). If in the course of that any of the `read`s return zero bytes (indicating EOF), the data is incomplete [1]. If after the loop an extra `read` after returns _not_ zero bytes, then there is unexpected extra data.

[1] I guess we could improve the `utxo_to_sqlite.py` with a `read()` wrapper that prints a nice "incomplete dump data" error message in that case, rather than continuing with an empty byte-string and likely running into an `AssertionError` or `IndexError` (or similar) at some point...

rkrux reviewed (commented):
ACK 509d871fc00e3dd5a9c42e267da13a322b646d76

> $ ./contrib/utxo-tools/utxo_to_sqlite.py /tmp/utxo_fifo ./utxo.dat

In PR description, should it not be `s/utxo.dat/utxo.sqlite` as it is the `outfile`?

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -3167,7 +3169,7 @@ static RPCHelpMan dumptxoutset()
                                         temppath,
                                         node.rpc_interruption_point);
-    fs::rename(temppath, path);
+    if (!fs::is_fifo(path_info)) fs::rename(temppath, path);
```
nit: Could use brackets

fjahr reviewed (commented):
utACK 509d871fc00e3dd5a9c42e267da13a322b646d76

Agree with prior comments that `utxo_to_sqlite.py` should probably just use this automatically in a follow-up or otherwise it would be good to add a hint in the documentation.

sedited commented:
This could also get a small release note. I'm sure we'd re-ACK quickly if you also want to address the outstanding nits.

theStack committed:
doc: add release note for #31560 (named pipe support for `dumptxoutset` RPC)

sedited reviewed (approved):
Re-ACK b19caeea098f92a7f72aaeee49573358f4b153a3

fjahr reviewed (commented):
utACK b19caeea098f92a7f72aaeee49573358f4b153a3

EDIT: Post-merge I guess ^^
